core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/KeyLock.scala

```diff
@@ -0,0 +1,79 @@
+package org.broadinstitute.dsde.workbench.leonardo
+
+import java.util.concurrent.TimeUnit
+
+import cats.effect.concurrent.{MVar, MVar2}
+import cats.effect.implicits._
+import cats.effect.{Blocker, Concurrent, ConcurrentEffect, ContextShift, Effect}
+import cats.implicits._
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+
+import scala.concurrent.duration.FiniteDuration
+
+abstract class Lock[F[_]] {
+  def acquire: F[Unit]
+  def release: F[Unit]
+  def withLock[A](fa: F[A]): F[A]
+}
+object Lock {
+  def apply[F[_]: Concurrent]: F[Lock[F]] =
+    MVar.of(()).map(l => new LockImpl(l))
+
+  // A Lock implementation backed by an MVar
+  // Inspired from https://typelevel.org/cats-effect/concurrency/mvar.html#use-case-asynchronous-lock-binary-semaphore-mutex
+  final private class LockImpl[F[_]: Concurrent](lock: MVar2[F, Unit]) extends Lock[F] {
```

Comment:
with Semaphore, you can have permit to require and release, is that not sufficient?

Reply:
Thinking why I didn't use `Semaphore`. :) I think since we just want a mutex then `Semaphore` is a bit too powerful. But yet `Semaphore(1)` would be equivalent -- could try to use that.